File: DriftGame/uracer-desktop/src/com/bitfire/uracer/game/logic/gametasks/PhysicsStep.java
package com.bitfire.uracer.game.logic.gametasks;

import com.badlogic.gdx.physics.box2d.World;
import com.bitfire.uracer.configuration.PhysicsUtils;
import com.bitfire.uracer.game.GameEvents;
import com.bitfire.uracer.game.events.PhysicsStepEvent.Type;
import com.bitfire.uracer.game.events.TaskManagerEvent;

public class PhysicsStep extends GameTask {
    private final World world;

    public PhysicsStep(World world, TaskManagerEvent.Order order) {
        super(order);
        this.world = world;
    }

    @Override
    public void dispose() {
        super.dispose();
        GameEvents.physicsStep.removeAllListeners();
    }

    @Override
    protected void onTick() {
        GameEvents.physicsStep.trigger(this, Type.OnBeforeTimestep);
        world.step(PhysicsUtils.Dt, 10, 10);
        GameEvents.physicsStep.trigger(this, Type.OnAfterTimestep);
    }

    @Override
    protected void onTickCompleted() {
        world.clearForces();
        GameEvents.physicsStep.trigger(this, Type.OnSubstepCompleted);
    }

    @Override
    public void onGameReset() {
        world.clearForces();
    }

    @Override
    public void onGameRestart() {
        onGameReset();
    }
}

File: DriftGame/uracer-desktop/src/com/bitfire/uracer/game/logic/gametasks/PhysicsStep.kt
package com.bitfire.uracer.game.logic.gametasks

import com.badlogic.gdx.physics.box2d.World
import com.bitfire.uracer.configuration.PhysicsUtils
import com.bitfire.uracer.game.GameEvents
import com.bitfire.uracer.game.events.PhysicsStepEvent
import com.bitfire.uracer.game.events.TaskManagerEvent

class PhysicsStep(private val world: World, order: TaskManagerEvent.Order) : GameTask(order) {

    override fun dispose() {
        super.dispose()
        GameEvents.physicsStep.removeAllListeners()
    }

    override fun onTick() {
        GameEvents.physicsStep.trigger(this, PhysicsStepEvent.Type.OnBeforeTimestep)
        world.step(PhysicsUtils.Dt, 10, 10)
        GameEvents.physicsStep.trigger(this, PhysicsStepEvent.Type.OnAfterTimestep)
    }

    override fun onTickCompleted() {
        world.clearForces()
        GameEvents.physicsStep.trigger(this, PhysicsStepEvent.Type.OnSubstepCompleted)
    }

    override fun onGameReset() {
        world.clearForces()
    }

    override fun onGameRestart() {
        onGameReset()
    }
}
